vortex-btrblocks/src/integer/dictionary.rs

```diff
@@ -0,0 +1,143 @@
+//! Dictionary compressor that reuses the unique values in the `IntegerStats`.
+
+use vortex_array::array::PrimitiveArray;
+use vortex_array::validity::Validity;
+use vortex_array::IntoArray;
+use vortex_buffer::Buffer;
+use vortex_dict::DictArray;
+use vortex_error::VortexResult;
+
+use crate::integer::stats::ErasedStats;
+use crate::integer::IntegerStats;
+
+macro_rules! typed_encode {
+    ($stats:ident, $typed:ident, $validity:ident, $typ:ty) => {{
+        let values: Buffer<$typ> = $typed.distinct_values.keys().copied().collect();
+
+        let max_code = values.len();
+        let codes = if max_code <= u8::MAX as usize {
+            let buf =
+                <DictEncoder as Encode<$typ, u8>>::encode(&values, $stats.src.as_slice::<$typ>());
+            PrimitiveArray::new(buf, $validity.clone()).into_array()
+        } else if max_code <= u16::MAX as usize {
+            let buf =
+                <DictEncoder as Encode<$typ, u16>>::encode(&values, $stats.src.as_slice::<$typ>());
+            PrimitiveArray::new(buf, $validity.clone()).into_array()
+        } else {
+            let buf =
+                <DictEncoder as Encode<$typ, u32>>::encode(&values, $stats.src.as_slice::<$typ>());
+            PrimitiveArray::new(buf, $validity.clone()).into_array()
+        };
```

Comment:
this effectively inlines the downscaling of the dictionary codes, so that we use the smallest type that fits the max code